pommerman/network/ion_server/__init__.py

```diff
@@ -0,0 +1,130 @@
+        port,
+        timeout,
+        mode,
+        ui_en=True,
+        exit_handler=True):
+    """Arguments:  
+    * max_players: The maximum amount of concurrent players  
+    * max_matches: The maximum amount of concurrent matches (Usually this \
+should be int(max_players/4))
+    * port: The port used by the server  
+    * timeout: The time to wait before issuing the STOP action  
+    * mode: The flavor of pommerman  
+    * ui_en: If True, UI is enabled else UI is disabled  
+    * exit_handler: If True, the exit handler is set else the exit handler \
+isn't set  
+    Description: This function is responsible for running the server."""
+    netpipe, rnetpipe = multiprocessing.Pipe()
+    netqueue = multiprocessing.Queue()
+    sp_net = multiprocessing.Process(
+        target=network.thread,
+        args=(rnetpipe, netqueue, port, max_players, mode, timeout))
+    sp_net.start()
+    if exit_handler:
+        signal.signal(signal.SIGINT, _exit_handler(sp_net))
+    if ui_en:
+        ui.info(ui.yellow, Strings.server_ready.value, ui.white,
+                ui.Symbol("✔", ":)"))
+    while (True):
+        netpipe.send([constants.SubprocessCommands.get_players.value])
+        players = netpipe.recv()
+        if players[2] < max_matches:
+            for x in list(players[0]["room"].keys()):
+                i = players[0]["room"][x]
+                if len(i) >= 4:
+                    MATCHSP.append(_create_match(i, netqueue, mode))
+                    del players[0]["room"][x]
+                    netpipe.send([
+                        constants.SubprocessCommands.update_cc.value, players[0]
+                    ])
+            if len(players[0]["noroom"]) >= 4:
+                e = random.sample(players[0]["noroom"],
+                                  (int(len(players[0]["noroom"]) / 4) * 4))
+                for group in range(int(len(players[0]["noroom"]) / 4)):
+                    MATCHSP.append(
+                        _create_match(e[group * 4:(group + 1) * 4], netqueue,
+                                      mode))
+                    for player in e[group * 4:(group + 1) * 4]:
+                        del players[0]["noroom"][players[0]["noroom"].index(
+                            player)]
+                    netpipe.send([
+                        constants.SubprocessCommands.update_cc.value, players[0]
+                    ])
+        if ui_en:
+            ui.info(ui.white, Strings.server_players.value, ui.yellow, "[",
+                    players[1], "/", max_players, "]", ui.white,
+                    Strings.server_matches.value, ui.yellow, "[", players[2],
+                    "/", max_matches, "]")
+        time.sleep(2)
+
+
+def _create_match(players, queue_subproc, mode):
+    """Description: This function is responsible for creating a match"""
+    subprocess = multiprocessing.Process(
+        target=match.thread, args=(players, queue_subproc, mode))
+    subprocess.start()
+    return subprocess
```

Comment:
Is there a cap on the number of subprocesses? It doesn't appear that we keep track of those other than in MATCHSP. Can you make sure that the processes within MATCHSP close? Otherwise, you may think that you are ending matches but the subprocesses are still open.

Reply:
I agree. This is something I was going to bring up on Discord. Actually, the sub-processes now do explicitly call `exit(0)` after they are done working. However, I was wondering how should I remove them from the MSP array after that..

Reply:
Actually, I just fixed this. I set the name of the process to "delete" at the end and added a part to the main thread that'll recursively check the names of the threads to match "delete" and will kill and remove those threads.

Reply:
Thank you! Please leave this unresolved and I'll return to it when I'm more cogent.